### PR TITLE
fix(setup): guard against pending-reboot indicators before Boxstarter runs

### LINKED ISSUE #149: fix(setup): 保留中の再起動指標が残ると Boxstarter が無限再起動ループに陥る
## Background

`setup.cmd` は Boxstarter を通じて `boxstarter.ps1`（Phase 0〜7 の実際の
セットアップ処理）を実行するが、Boxstarter 自身の `choco install` ラッパー
（`Boxstarter.Chocolatey\Chocolatey.ps1` の `chocolatey` 関数、およそ212行目）
は、実際にインストール処理を呼び出す**前**に `Test-PendingReboot`
（`Get-PendingReboot`）を評価し、真であれば `boxstarter.ps1` を一度も
実行せずに即座に `Invoke-Reboot` して return する:

```powershell
if (((Test-PendingReboot) -or $Boxstarter.IsRebooting) -and $Boxstarter.RebootOk) {
    return Invoke-Reboot   # ここで即再起動。boxstarter.ps1 の Phase 0〜7 は一度も走らない
}
```

`Get-PendingReboot` が「再起動保留あり」と判定する材料の一つに
`HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager` の
`PendingFileRenameOperations` があるが、この関数自身のドキュメント
コメントが明記する通り、これは既知の誤検知源である
（"some Anti-Virus leverage this key for def/dat removal, giving a
false positive PendingReboot"）。

実機で次の事象を確認した: このレジストリ値に
`gamingservicesproxy_13.dll`（Gaming Services 関連の数週間前の更新の
残骸で、`.dll.0` という退避用ファイルが実際にディスク上に残ったまま）
を指す**ゾンビ状態のエントリ**が居座っており、通常の再起動では一切
解消されなかった。結果として:

- ログオン → Boxstarter のスタートアップフォルダ再開スクリプト
  (`boxstarter-post-restart.bat`) が `Invoke-Boxstarter` を再実行
- 同じ古いフラグを検知 → 即座に再起動
- 以降ログオンのたびに同じことを繰り返す

という無限ループに陥った。Boxstarter にはこの状態から自力で抜け出す
ための試行回数の上限機構が存在しない。

実機のログで検証済み: `%LOCALAPPDATA%\Boxstarter\boxstarter.log` および
`C:\ProgramData\chocolatey\logs\chocolatey.log` を確認したところ、
13:27〜14:14 の間に13回のループが発生したが、いずれの回でも
`choco install`（＝ `boxstarter.ps1` の実行）は一度も呼ばれておらず、
`boxstarter.ps1` の `[Phase N]` 出力も一切記録されていない。つまり
**本リポジトリの `boxstarter.ps1` 自身の Phase 0〜7 のロジックには
問題がなく**、その手前（Boxstarter 本体の再起動判定）で永久に
足止めされていたことが確認できている。

さらに、Boxstarter は最初の再起動前に UAC を無効化し
(`EnableLUA=0`)、自動ログオンを設定する
(`AutoAdminLogon=1` + 資格情報を `%LOCALAPPDATA%\Boxstarter\Boxstarter.autologon`
にキャッシュ)。両方とも、`boxstarter.ps1` の Phase 7
「Teardown」（`Enable-UAC` 呼び出し）まで到達して初めて復元される。
ループが Phase 0 にすら到達しない今回のケースでは、UAC 無効化・
自動ログオン設定・スタートアップの再開スクリプトがすべて残置された
ままとなり、Windows Update や電源断など setup とは無関係な理由で
次に再起動が起きた場合、無音のまま同じループが再開してしまう
(UAC が無効なままで)。実機では手動でこれらすべてを是正済み
(`PendingFileRenameOperations` クリア、`AutoAdminLogon` 解除、
`EnableLUA` を 1 に復元、`boxstarter-post-restart.bat` 削除) だが、
setup.windows / Boxstarter のどちらにも、これを自動検知・復旧する
仕組みが存在しない。

この issue が対象とするのは、あくまで setup.windows 側から現実的に
実装可能な緩和策である。Boxstarter 本体の「再起動判定が晴れない場合の
打ち切り機構の欠如」自体は upstream (`chocolatey/boxstarter`) 側の
制約であり、`boxstarter.ps1`（本リポジトリが所有するコード）は
Boxstarter に制御が渡った**後**の再開サイクルには一切関与できない
(再開はスタートアップフォルダのバッチ経由で Boxstarter 自身が完結して
行うため、setup.windows のコードが再実行される余地がない)。したがって
「再起動回数の上限カウンタ」を `boxstarter.ps1` 内に実装しても、
今回のケースでは一度も呼ばれずに意味を持たない。setup.windows が
確実に関与できる唯一のタイミングは、**最初に `Install-BoxstarterPackage`
を呼び出す直前**の一度きりの事前チェックである。

## Proposed change

`setup.cmd` から `Install-BoxstarterPackage` を呼び出す直前に、
Boxstarter が参照するのと同種の再起動保留指標を事前チェックする
ガードを追加する。

1. 新規 `libs/reboot-guard.ps1` を追加し、`Test-OsSupport`
   (`libs/os-guard.ps1`) と同様のスタイルで、以下を検査して結果を
   返す関数（例: `Test-PendingRebootIndicators`）を実装する:
   - `HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Component Based Servicing\RebootPending`
     の存在
   - `HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\RebootRequired`
     の存在
   - `HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager` の
     `PendingFileRenameOperations` 値の有無
   - コンピューター名変更保留 (`HKLM:\SYSTEM\CurrentControlSet\Control\ComputerName\ActiveComputerName`
     と `\ComputerName` の不一致、または
     `HKLM:\SYSTEM\CurrentControlSet\Services\Netlogon\JoinDomain`
     / `AvoidSpnSet` の保留)
   - 戻り値には、どの指標が立っていたかを個別に判別できる形
     (例: 各指標名をキーとする Hashtable、または
     `[PSCustomObject]` のプロパティ) を持たせる。
     `Get-PendingReboot`（Boxstarter.Bootstrapper）と同じ判定材料を
     参照する設計にするが、この関数自体は Boxstarter モジュールに
     依存せず、レジストリを直接読む。
2. `setup.cmd` の `Installing Chocolatey...` より前、
   `Install-BoxstarterPackage` を呼び出す直前に、1. のチェックを
   PowerShell 経由で呼び出す一行を追加する。いずれかの指標が
   立っていた場合:
   - どの指標が立っているかを名指しした警告メッセージを表示する
   - 「これが本物の保留中再起動なら通常の再起動で解消するが、
     解消しない場合は既知の誤検知(特に `PendingFileRenameOperations`)
     であり、そのまま続行すると Boxstarter が無限に再起動を繰り返す
     可能性がある」旨と、調査・復旧手順（本 issue の
     Proposed change 3. で追加するドキュメントへのリンク）を案内する
   - **既定ではここで中断する**(非ゼロ終了)。対話的な確認プロンプトは
     追加しない(setup.cmd は非対話実行される経路もあり得るため)。
     ユーザーが状況を把握した上で明示的に続行したい場合のための
     オプトアウト手段(例: 環境変数
     `SETUP_IGNORE_PENDING_REBOOT=1`、または `setup.cmd` の追加引数)
     を用意し、その使い方も警告メッセージ内に案内する。
   - いずれの指標も立っていない場合は、現状と完全に同じ挙動
     (無警告で `Install-BoxstarterPackage` を呼び出す) を維持する。
3. `README.md`（および `README.ja.md` があれば同様に）へ
   "Troubleshooting" セクションを追加し、次を記載する:
   - 症状: `setup.cmd` を実行すると再起動を繰り返し、
     `[Phase 0]` 以降のコンソール出力が一度も表示されない
   - 原因: 古い/誤検知の再起動保留指標(典型例は
     `PendingFileRenameOperations`)が残っており、通常の再起動では
     解消しない
   - 調査手順: 該当レジストリ値を確認する PowerShell スニペット
   - 復旧手順: 該当する古いエントリの削除、および
     `EnableLUA`(UAC)・`AutoAdminLogon`・
     `%APPDATA%\Microsoft\Windows\Start Menu\Programs\Startup\boxstarter-post-restart.bat`
     の状態確認と復旧(ループに一度でも陥っていた場合、これらは
     Boxstarter によって変更されたまま残っている可能性があるため)

## Acceptance criteria

- `tests/powershell/reboot-guard.Tests.ps1` を新規追加し、
  `libs/os-guard.ps1` 系のテストと同様に `Test-PendingRebootIndicators`
  を Pester でカバーする:
  - 4種の指標がいずれも立っていない場合、指標なしと判定されること
  - 4種の指標を個別にモックし、それぞれ単独で立っている場合に
    正しく名指しで検出されること
  - 複数の指標が同時に立っている場合、すべてが検出結果に含まれること
- `setup.cmd` の変更について、指標をモック/一時設定した状態と
  していない状態の両方で、想定どおりの標準出力・終了コード
  (指標なし: 現状と同じ挙動で `Install-BoxstarterPackage` を呼び出す/
  指標あり: 警告と指標名を表示して非ゼロで終了し、
  `Install-BoxstarterPackage` を呼び出さない/
  オプトアウト指定時: 警告を表示しつつ従来どおり続行する)
  になることを手動で確認する
  (`setup.cmd` はバッチファイルのため Pester の対象外)
- PSScriptAnalyzer が `libs/reboot-guard.ps1` に対して指摘を出さない
- `pre-push-validate` 相当のフルチェーンがクリーンな作業ツリーに
  対して exit 0 で完了する
- `README.md` に Troubleshooting セクションが追加され、
  markdownlint / cspell を通過する

## IDD ブートストラップ

marker prefix は `setup-windows`。実行フローは
[`docs/idd-workflow.md`](docs/idd-workflow.md) を参照。

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

_Effort: M -- author-estimated size; a soft autopilot
selection tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Boxstarter's own pre-flight reboot check (`Get-PendingReboot`, evaluated
before `boxstarter.ps1` ever runs) can get stuck on stale or
false-positive registry indicators -- most notably
`PendingFileRenameOperations`, a documented false-positive source -- and
silently reboot the machine forever without ever invoking this repo's
setup logic.

This PR adds a pre-check that runs before `Install-BoxstarterPackage` is
ever called:

- `libs/reboot-guard.ps1` (`Test-PendingRebootIndicators`): inspects the
  same class of indicators Boxstarter itself references
  (`ComponentBasedServicing`, `WindowsUpdateAutoUpdate`,
  `PendingFileRenameOperations`, and a `ComputerRenamePending` composite
  covering computer-name-mismatch and pending Netlogon domain-join
  keys), without depending on the Boxstarter module.
- `setup.cmd`: calls the check immediately before
  `Install-BoxstarterPackage` (after the Chocolatey/Boxstarter bootstrap
  steps, so a reboot-pending state those steps leave behind is still
  caught). If any indicator is set, it warns which one(s), points to the
  new Troubleshooting section, then waits 30 seconds (`timeout /t 30`,
  so a double-clicked `setup.cmd` doesn't flash-close before the warning
  is readable) and aborts -- unless `SETUP_IGNORE_PENDING_REBOOT=1` is
  set, in which case it warns but continues as before. If nothing is
  set, behavior is unchanged (silent, continues).
- `README.md` / `README.ja.md`: a new Troubleshooting section covering
  the symptom, cause, an investigation snippet for all 4 indicators, and
  recovery steps (clearing the stale registry value, restoring
  `EnableLUA`/`AutoAdminLogon`, removing a leftover
  `boxstarter-post-restart.bat`) for the case where a machine already
  got stuck in the loop before this guard existed.
- `tests/powershell/reboot-guard.Tests.ps1`: Pester coverage for
  `Test-PendingRebootIndicators`, mocking `Test-Path`/`Get-ItemProperty`
  (the `HKLM:` drive doesn't exist on the Linux CI runner) with
  per-path `-ParameterFilter`s and `Should -Invoke` assertions proving
  each indicator actually probes its own registry path, not just that
  the aggregate return value happens to match.

Closes #149

## Manual verification of `setup.cmd`'s three paths

`setup.cmd` is a batch file, so it's outside Pester's reach (per the
issue's own acceptance criteria, this is a manual check). I extracted
the exact PowerShell one-liner now embedded in `setup.cmd` into a
standalone `.cmd` file and ran all three scenarios through real
`cmd.exe`, with `libs/reboot-guard.ps1` swapped for a stub that forces
specific indicators (to avoid touching real `HKLM` values):

- **No indicator set**: exit 0, no warning, falls through to
  `Install-BoxstarterPackage` unchanged.
- **Indicator(s) set, no opt-out**: prints the named indicator(s) and
  guidance, `errorlevel 1` is set, `setup.cmd`'s abort block fires
  (`timeout /t 30` + `exit /b 1`), `Install-BoxstarterPackage` is never
  reached.
- **Indicator(s) set + `SETUP_IGNORE_PENDING_REBOOT=1`**: prints the
  same warnings plus a "continuing" notice, exits 0, falls through as
  before.

I also confirmed `Test-PendingRebootIndicators` runs cleanly against the
real registry under both `pwsh` and `powershell.exe` 5.1 (the actual
host `setup.cmd` invokes) on this machine, with no pending-reboot
indicators currently set here.

## Design note: the `timeout /t 30` on abort

The issue's own proposed change says not to add an interactive
confirmation prompt, since `setup.cmd` can run non-interactively. The
abort path doesn't use `pause` -- it uses a bounded `timeout /t 30`
instead, so the exit code is always 1 and a keypress only skips the
wait early rather than gating it. This exists only so a double-clicked
`setup.cmd` (the common case this bug report describes) doesn't close
before the warning can be read, without risking an indefinite hang if
this runs unattended with a console still attached but nobody there to
press a key. A fully unattended/non-interactive invocation piping
stdin (or redirected from `NUL`) will not block on it either way.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01S9TCtbMGY6nAmB2vUvadVM

## Summary by CodeRabbit

* **New Features**
  * Setup now detects pending Windows reboots before installing required components.
  * Setup warns and stops by default when a pending reboot is detected, with an option to continue using `SETUP_IGNORE_PENDING_REBOOT=1`.
  * Added troubleshooting guidance for diagnosing and recovering from repeated setup restarts.

* **Tests**
  * Added coverage for common Windows reboot indicators and edge cases.

### CODE REVIEW

File: libs/reboot-guard.ps1 [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  .OUTPUTS
+  PSCustomObject with one boolean property per indicator:
+  ComponentBasedServicing, WindowsUpdateAutoUpdate,
+  PendingFileRenameOperations, ComputerRenamePending.
+  #>
+  $componentBasedServicing = Test-Path 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Component Based Servicing\RebootPending'
+  $windowsUpdateAutoUpdate = Test-Path 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\WindowsUpdate\Auto Update\RebootRequired'
+
+  $pendingFileRenameProps = Get-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager' -Name PendingFileRenameOperations -ErrorAction SilentlyContinue
+  $pendingFileRenameOperations = [bool]($pendingFileRenameProps -and $pendingFileRenameProps.PendingFileRenameOperations)
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply libs/reboot-guard.ps1]: **Accepted** — fixed in c9a9a40: filtered PendingFileRenameOperations entries to non-empty/non-whitespace before casting to bool, so an array with 2+ all-empty entries (which PowerShell always reads as truthy regardless of content) no longer misreads as pending. The genuine [source, empty-dest] pending shape still detects correctly. Added Pester coverage for both cases.

File: tests/powershell/reboot-guard.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    $result.ComputerRenamePending | Should -BeFalse
+    Should -Invoke Get-ItemProperty -Times 1 -ParameterFilter {
+      $Path -eq 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager' -and $Name -eq 'PendingFileRenameOperations'
+    }
+  }
+
+  It 'treats a single-empty-string PendingFileRenameOperations value as not pending' {
+    # Boxstarter's own Get-PendingReboot has the same blind spot: a
+    # REG_MULTI_SZ holding one empty string is falsy via plain `if
+    # ($value)`, matching the check this function uses.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/powershell/reboot-guard.Tests.ps1]: **Rejected** — verified empirically: [bool]@('') is $false in PowerShell (a 1-element array unwraps to its single element's own truthiness, and an empty string is falsy), so the original comment's claim was correct as written. The array-truthiness concern IS real for a 2+-element all-empty array though ([bool]@('','') is $true) -- filtered for that case in c9a9a40, and rewrote the comment to describe the intentional filter rather than the now-unused incidental array-unwrap framing.

File: setup.cmd [hunk 3]
```diff
@@ -9,6 +9,18 @@ echo.
 REM --- Unblock all PowerShell scripts ---
 powershell -Command "Get-ChildItem -Recurse '%~dp0libs\*.ps1' | Unblock-File"
 
+REM --- Check for pending-reboot indicators before installing anything ---
+REM (a stale/false-positive indicator here can otherwise make Boxstarter
+REM  loop rebooting forever without ever running this repo's setup logic)
+powershell -NoProfile -ExecutionPolicy Bypass -Command ^
+  ". '%~dp0libs\reboot-guard.ps1'; $r = Test-PendingRebootIndicators; $flagged = $r.PSObject.Properties.Name | Where-Object { $r.$_ }; if ($flagged) { Write-Warning ('Pending-reboot indicator(s) detected: ' + ($flagged -join ', ')); Write-Warning 'This may be a real pending reboot (reboot and re-run), or a known false positive -- see README.md Troubleshooting.'; Write-Warning 'If unresolved, Boxstarter can loop rebooting forever without ever running this repo''s setup logic.'; Write-Warning 'To proceed anyway, set SETUP_IGNORE_PENDING_REBOOT=1 and re-run setup.cmd.'; if ($env:SETUP_IGNORE_PENDING_REBOOT -ne '1') { exit 1 } else { Write-Warning 'SETUP_IGNORE_PENDING_REBOOT=1 is set; continuing despite the indicator(s) above.' } }"
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply setup.cmd]: **Accepted** — fixed in f8c014e: moved the guard block to immediately before Install-BoxstarterPackage, after the Chocolatey/Boxstarter bootstrap steps, so a pending-reboot indicator either of those leaves behind is still caught. Updated README.md/README.ja.md to describe the new position.

**Comment by kurone-kito** [comment]
**Accepted** — fixed in f8c014e: the same "run the guard earlier" risk this walkthrough's Merge Risk note flags is addressed by moving the guard block to immediately before `Install-BoxstarterPackage`, after the Chocolatey/Boxstarter bootstrap steps. See the disposition on the corresponding inline thread (https://github.com/kurone-kito/setup.windows/pull/150#discussion_r3931819541) for detail.

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector did not review HEAD f8c014e (or any earlier HEAD on this PR); its comment reports that its review usage allowance was exhausted, so this is not a completed review (source: https://github.com/kurone-kito/setup.windows/pull/150#issuecomment-5537116780).

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-code-16272ff9 f8c014e545b3bf279b5e575e0e4d66bbbd7e1fdd 2026-09-04T07:43:43Z

File: setup.cmd [hunk 4]
```diff
@@ -21,6 +21,21 @@ REM --- Ensure Boxstarter is available ---
...
+REM --- guards, so a reboot-pending state left behind by the Chocolatey
+REM --- or Boxstarter install steps above is still caught.
+powershell -NoProfile -ExecutionPolicy Bypass -Command ^
+  ". '%~dp0libs\reboot-guard.ps1'; $r = Test-PendingRebootIndicators; $flagged = $r.PSObject.Properties.Name | Where-Object { $r.$_ }; if ($flagged) { Write-Warning ('Pending-reboot indicator(s) detected: ' + ($flagged -join ', ')); Write-Warning 'This may be a real pending reboot (reboot and re-run), or a known false positive -- see README.md Troubleshooting.'; Write-Warning 'If unresolved, Boxstarter can loop rebooting forever without ever running this repo''s setup logic.'; Write-Warning 'To proceed anyway, set SETUP_IGNORE_PENDING_REBOOT=1 and re-run setup.cmd.'; if ($env:SETUP_IGNORE_PENDING_REBOOT -ne '1') { exit 1 } else { Write-Warning 'SETUP_IGNORE_PENDING_REBOOT=1 is set; continuing despite the indicator(s) above.' } }"
+if errorlevel 1 (
+  echo.
+  echo Aborting: unresolved pending-reboot indicator detected. See README.md Troubleshooting section.
+  pause
+  exit /b 1
+)
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply setup.cmd:41]: **Accepted** — fixed in 47f8725: replaced pause with timeout /t 30 on the abort path. Verified empirically: with stdin redirected from NUL (fully automated), timeout still exits immediately without hanging (prints a harmless "Input redirection is not supported" notice, same as before); with a console attached but unattended, timeout now proceeds after 30s instead of blocking forever like pause did.

**Comment by kurone-kito** [comment]
review-ack: claude-code-16272ff9 47f87250bb0b2f4648d80c39f0b83484c641eec3 2026-09-04T07:51:46Z

**Comment by kurone-kito** [comment]
| Field | Value |
| --- | --- |
| Phase | F3 merged |
| Claim | d630ee6b-96ef-4349-9558-b306609f7b2d |
| Branch | issue/149-fix-setup-boxstarter |
| Last checked | 2026-09-04T08:18:29.027Z |
| Open blockers | none |
| Next action | F4 cleanup then F5 discover |
| Authoritative by | merge commit ff1dcbe0 (matched head 47f87250) |

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-code-16272ff9 47f87250bb0b2f4648d80c39f0b83484c641eec3 2026-09-04T07:52:19Z

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai did not review HEAD 47f8725 (or any HEAD after e43da30); this comment was edited in place to report that its review allowance was exhausted, so this is not a completed review (source: https://github.com/kurone-kito/setup.windows/pull/150#issuecomment-5537136727). The walkthrough content still shown above it is stale leftover from the e43da30 review — its `final_review_risk_coverage` metadata still cites `e43da308` as the covered commit, and its Merge Risk note repeats the same guard-ordering concern already fixed in f8c014e and confirmed on the corresponding inline thread (https://github.com/kurone-kito/setup.windows/pull/150#discussion_r3931819541). Not a fresh finding.

**Comment by kurone-kito** [comment]
_claude-code-16272ff9: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-code-16272ff9: critique baseline — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**F4 Cleanup Evidence**

| Field                            | Value                                                            |
| --------------------------------- | ----------------------------------------------------- |
| Status                           | applied                                                          |
| Applied                          | 11                                                                |
| Failed                           | 0                                                                  |
| Skipped                          | 6                                                                  |
| Permission-blocked               | 0                                                                  |
| Retry attempts (bound-exhausted) | 1 (false)                                                    |
| Notes                            | Skipped items are informational, non-blocking: 1 known review-bot regular comment lacks a completed-review signal, 2 review threads + 3 review parents are missing an IDD accept/reject disposition marker or have no associated threads. |
